Improve how already-downloaded plugin extensions are handled.

amisevsk opened the pull request:
### What does this PR do?
Allows caching of installed plugin extensions in the workspace. Instead of erasing the entire `/plugins` directory on every start, the broker will instead store a json log of installed plugins, and use this to only download required extensions, via the rule:

1. If a plugin is installed locally but not in the current workspace, all of its artifacts are removed
2. If a plugin is not installed locally and included in the current workspace, all of its artifacts will be downloaded
3. If a plugin is installed but the artifacts to be downloaded are different compared to a prior run (e.g. a new version of `latest`), then 
    - Extensions that are no longer needed are removed
    - New extensions are downloaded
    - Extensions that are shared between the old and new version are reused.

These changes make the assumption that all extension artifacts are served from unique URLs, and will fail if e.g. the `.vsix` file at a URL is updated.

Installed plugin metadata is stored in a new struct (see `model.InstalledPlugin`), which stores the plugin ID, whether the plugin is remote (runs in a sidecar), and a map of extension URLs to filesystem paths. A list of installed plugins is saved in the `/plugins` directory along with an internal file version number to allow versioned changes.

This PR significantly improves brokering time for the most common use case in starting a workspace, where a previously started non-ephemeral workspace is started a second time; in this case brokering is complete almost immediately, since nothing needs to be downloaded.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/15470

sleshchenko commented on brokers/artifacts/extensions.go:
```diff
@@ -0,0 +1,148 @@
+		if match == nil {
+			// Plugin has been uninstalled since last start
+			b.PrintInfo("Uninstalling plugin: %s", plugin.ID)
+			if err := b.removePlugin(plugin); err != nil {
```
It was quite unexpected for me that `getPlan` method actually removes plugins, maybe instead of performing any operation here it makes sense to keep it only analyzing extenstions and return plan:
```
func (b *Broker) getPlan(requested, installed []model.CachedPlugin) (toInstall []model.CachedPlugin, toRemove []model.CachedPlugin, filesToRemove []File) {
```
and then remove them on another level? Or maybe there can be found a more clear method name.

P.S. I have not had reviewed your changes deeply enough if my comment does not make any sense - please just mark it resolved =)

amisevsk replied:
You're right that the name is not very good, I will improve that. The issue I ran into with trying to get a plan and then execute it is that there are three cases:
1. A plugin is added to the workspace -> download its extensions
2. A plugin is removed from the workspace -> remove its extensions
3. A plugin is updated in the workspace -> remove outdated extensions, download new ones

The third case is hard to support and also potentially the most common (`latest` represents an "update", sort of). What I opted for, as a result, is to partially fill plugins in the plan to support case three (if an extension has already been downloaded, it has the extension's path added to the struct).

Looking back this is probably over-engineered, since it only occurs when e.g. jdt is updated but the debug extension is not.

nickboldt replied:
What about scenarios where the same extension (but different versions) are installed because of plugins which depend on different versions? eg., 
```
redhat/java8/latest 
- https://download.jboss.org/jbosstools/vscode/3rdparty/vscode-java-debug/vscode-java-debug-0.20.0.vsix

redhat/quarkus-java8
- https://download.jboss.org/jbosstools/vscode/3rdparty/vscode-java-debug/vscode-java-debug-0.23.0.vsix
```

Do you make sure that an update for one plugin (eg., newer quarkus results in 0.23.0 -> 0.24.0) doesn't also garbage collect the 0.2.0 used by the other plugin?

amisevsk replied:
Currently, each plugin gets its own .vsix on its own path, so there's no overlap for that case. E.g. redhat/java plugins end up in
```
/plugins
└── sidecars
    └── redhat_java_0_43_0
        ├── redhat.java.0.43.0.kmwjfvvtpc.java-0.43.0-1473.vsix
        └── redhat.java.0.43.0.wgfmdnrsmo.vscode-java-debug-0.16.0.vsix
```

sleshchenko commented on brokers/artifacts/broker.go:
```diff
@@ -58,122 +57,125 @@ func (b *Broker) Start(pluginFQNs []model.PluginFQN, defaultRegistry string) err
-		err = b.ProcessPlugin(meta)
+	requestedPlugins := convertMetasToPlugins(pluginMetas)
+
+	toInstall := b.SyncPlugins(requestedPlugins)
```
It makes an impression that not every plugin that is requested should be installed.
But later I found 
```go
func (b *Broker) getPlan(requested, installed []model.CachedPlugin) (toInstall []model.CachedPlugin) {
	toInstall = requested
        ...
```
Maybe then `b.SyncPlugins(requestedPlugins)` should return nothing...
Or consumes Metas and returns `CachedPlugin` ?

amisevsk replied:
The idea behind `SyncPlugins` is that it syncs the list of requested plugins with what's on disk, i.e.
- Requested plugins should contain already-installed extensions
- Disk should not contain plugins that are not requested

I think the unclear thing is that the convention for `model.CachedPlugin` is that if a extension already has a path, it doesn't need to be downloaded. We set `toInstall = requested` since at the end of the day, all requested plugins must be installed, and then we partially fill `toInstall`'s extension urls with already downloaded paths.

If we make `b.SyncPlugins(requestedPlugins)` return nothing, then we basically move all the functionality in `broker.go` into `extensions.go`. Currently the division of labour is 
- extensions.go: keep up local disk cache of plugins
- broker.go: download extensions when necessary

As an example, suppose a workspace has `redhat/vscode-openshift-connector/latest` installed, and a Che update has caused `latest` to go from `0.0.21` to `0.1.0`. Running `SyncPlugins` will return
```go
model.CachedPlugin{
  ID: redhat/vscode-openshift-connector/latest
  IsRemote: true
  CachedExtensions: map[string]string {
    "https://github.com/[...]/vscode-kubernetes-tools-1.0.0.vsix": "<path to downloaded vsix>"
    "https://github.com/[...]/redhat.vscode-yaml-0.4.0.vsix": "<path to downlaoded vsix>"
    "https://github.com/[...]/openshift-connector-0.0.21-228.vsix": "" // empty string means should be downloaded
  }
}
```

Then `ProcessPlugin` will only download `https://github.com/[...]/openshift-connector-0.0.21-228.vsix` and update that entry in the map.

amisevsk replied:
Renamed method to `SyncWithPluginsDir`, hopefully that makes things a little clearer.

sleshchenko replied:
> If we make b.SyncPlugins(requestedPlugins) return nothing, then we basically move all the functionality in broker.go into extensions.go. Currently the division of labour is
> 
> extensions.go: keep up local disk cache of plugins
> broker.go: download extensions when necessary

Just an idea to consider:

* broker.go: publish events, control downloading extensions and caching process by invoking the corresponding methods from files below
* extensions.go: download extensions when necessary
* cache.go: keep up local disk cache of plugins

Just minor comment what may be considered to be implemented in upcoming PRs.

amisevsk replied:
That makes a lot of sense, thanks.

sleshchenko commented on brokers/artifacts/broker.go:
```diff
@@ -58,122 +57,125 @@ func (b *Broker) Start(pluginFQNs []model.PluginFQN, defaultRegistry string) err
+// If a plugin already has artifacts downloaded for a given extension, that
+// extension is skipped.
+func (b *Broker) ProcessPlugin(plugin *model.CachedPlugin) error {
+	workDir, err := b.ioUtils.TempDir("", "vscode-extension-broker")
```
Just wondering if it would be better to rename the folder to `artifacts-broker` since there is no explicit VSCode related logic in code, everything in name in a general manner - Plugin, Extension.

amisevsk replied:
:+1: this is a holdover from before the refactor, will update.

amisevsk replied:
:+1:

sleshchenko replied:
> :+1: 

:+1: 

;)

sleshchenko commented on brokers/artifacts/broker.go:
```diff
@@ -58,122 +57,125 @@ func (b *Broker) Start(pluginFQNs []model.PluginFQN, defaultRegistry string) err
+		extensionIdx = extensionIdx + 1
+		b.PrintInfo("  Installing plugin extension %d/%d", extensionIdx, numExtensions)
+		if path != "" {
+			// Plugin has previously been downloaded, nothing needs to be done.
```
maybe it makes sense to print log that already downloaded archive is used

amisevsk replied:
:+1:

sleshchenko commented on brokers/artifacts/broker.go:
```diff
@@ -58,122 +57,125 @@ func (b *Broker) Start(pluginFQNs []model.PluginFQN, defaultRegistry string) err
+// relevant artifacts in their appropriate location in the /plugins directory.
+// If a plugin already has artifacts downloaded for a given extension, that
+// extension is skipped.
+func (b *Broker) ProcessPlugin(plugin *model.CachedPlugin) error {
```
AFAIU previously this method also generated configuration for Theia and VsCode plugins, but now it does not. Maybe it makes sense to rename this method to DownloadExtenstions?

amisevsk replied:
Done!

sleshchenko commented on brokers/artifacts/broker.go:
```diff
@@ -58,122 +57,125 @@ func (b *Broker) Start(pluginFQNs []model.PluginFQN, defaultRegistry string) err
+
+	for _, meta := range metas {
+		if !utils.IsTheiaOrVscodePlugin(meta) {
+			continue
```
In terms of Che Plugins model, even non-Theia or non-Vscode is a Che Plugin. Please consider putting here a comment why only Theia or VsCode plugins are returned here - like it's artifacts broker and only they have such...

sleshchenko commented on brokers/artifacts/extensions.go:
```diff
@@ -0,0 +1,148 @@
+	return pluginsJSON.Plugins, nil
+}
+
+func (b *Broker) writeInstalledPlugins(cached []model.CachedPlugin) error {
```
It's used outside of this file, maybe then it should be public as `SyncWithPluginsDir` is?

amisevsk commented:
@sleshchenko I've updated the PR (but forgot to split it into separate commits so it's hard to read -- sorry). Changes:
- Check if the error in reading installed_plugins.json is a "file does not exist error", and don't print the error message to the user in that case (https://github.com/eclipse/che-plugin-broker/pull/86#pullrequestreview-342463933)
- Reorganize the broker to split into files as you suggested (`broker.go`, `extensions.go`, and `cache.go`)
  - I reverted the name change `ProcessPlugin` -> `DownloadExtensions`, since that's technically a backwards-incompatible change from 3.0.0.
  - I also made the `cache.go` functions internal, since they really should be only used as part of the broker. The only exported functions in the broker are now `Start`, `PushEvents` and `ProcessPlugin`. None of the file syncing stuff is exported, since there's no way to easily expose a uniform interface -- one step has to happen before `DownloadExtensions` and the other has to happen after.

sleshchenko commented:
> I reverted the name change ProcessPlugin -> DownloadExtensions, since that's technically a backwards-incompatible change from 3.0.0.

Well, as far as I see `ProcessPlugin` comes from `common.BrokerImpl` that is not used anymore.
So, technically it's backward-incompatible but there are not much depending on Che Plugin Broker .
It seems a good time to review what we want to expose as public API of Plugin Broker... Is it `BrokerImpl`? If no - then I think we can afford such breaking change.
It definitely can be considered to implement later in the scope of the issue about making Plugin Broker usable as a library.

amisevsk commented:
In testing these changes on an actual cluster, I encountered a few issues that are fixed in the 3 new commits:
- 9d4f58b reworks how messages are pushed through the event bus, since logging multiple times quickly can cause messages to be out of order
- 5ca4b9c fixes a bug where removing a local plugin removed all remote plugins (due to erasing everything in `/plugins`)
- c4144e7 is a sanity check -- we ensure all cached extensions are present on disk before assuming we can use them

Scenarios tested:
- Ephemeral workspace (all plugins always downloaded)
- Non-ephemeral workspace
    - Add remote plugin to existing workspace (plugin is downloaded)
    - Add local plugin to existing workspace (plugin is downloaded)
    - Remove remote plugin (plugin extensions are removed, others untouched)
    - Remove local plugin (plugin extensions are removed, others untouched)
    - Delete `installed.json` -> `/plugins` erased and all plugins redownloaded
    - Delete an extension from an installed plugin -> `/plugins` erased and all plugins redownloaded.

There is still one outstanding issue, where the logs for `Saving log of installed plugins` is delivered before the final plugin is checked (I see it with java-maven + the vim extension).

@sleshchenko please take another look, if you have a moment.

amisevsk commented:
Opened issue https://github.com/eclipse/che/issues/15758 to address the out of order logs.

sleshchenko commented on brokers/artifacts/extensions.go:
```diff
@@ -0,0 +1,95 @@
+			infoLine.WriteString(fmt.Sprintf("\n    Plugin already downloaded"))
+			continue
+		}
+		infoLine.WriteString(fmt.Sprintf("\n    Downloading plugin from %s", URL))
```
With it, we get ordered output but we see nothing when something is happening, like downloading...

Maybe it would be better if we flush buffer each time before the time-consuming task is started?

like:
```go
        var infoLine strings.Builder
	defer b.PrintInfoBuffer(&infoLine)

	infoLine.WriteString(fmt.Sprintf("Processing plugin %s", plugin.ID))
	numExtensions := len(plugin.CachedExtensions)
	extensionIdx := 0
	for URL, path := range plugin.CachedExtensions {
		extensionIdx = extensionIdx + 1
		infoLine.WriteString(fmt.Sprintf("\n  Installing plugin extension %d/%d", extensionIdx, numExtensions))
		if path != "" {
			infoLine.WriteString(fmt.Sprintf("\n    Plugin already downloaded"))
			continue
		}
		infoLine.WriteString(fmt.Sprintf("\n    Downloading plugin from %s", URL))

                b.PrintInfoBuffer(&infoLine)
                infoLine.reset()
                ...
```
WDYT?

sleshchenko replied:
Another implementation I can imagine with channels, where one thread receives messages and collects in into buffer which is flushed like after some period, like 1-2 seconds.
But it would require even more changes in this PR, I'm OK to go with the current implementation.

amisevsk replied:
I agree that not showing anything while downloading is a problem, this is mostly a workaround to preserve logs in case of failure. With a slower internet connection, it does look strange, though. 

I like your suggestion of attempting to dump info more often, since as it's currently implemented the java plugin logs *both* extensions at once, which isn't ideal. The main goal is to keep the `downloading plugin x/#` logs for each extension together to avoid interleaving.

I've pushed a commit to rework this to log on every loop iteration and also move to using `[]string` in order to avoid worrying about newlines.

amisevsk commented on brokers/artifacts/broker.go:
```diff
@@ -58,122 +55,55 @@ func (b *Broker) Start(pluginFQNs []model.PluginFQN, defaultRegistry string) err
-	err = b.injectPlugin(meta, archivesPaths)
-	return err
+	b.PrintInfo("All plugin artifacts have been successfully downloaded")
+	b.PubDone("")
```
I think `b.PubDone()` doesn't display output, e.g. for the metadata broker it's used to pass the Che plugins json back to Che server. The empty `PubDone()` notifies the server that brokering completed successfully, but there is no tooling to return and Che server is not listening for it (I'm not certain Che server even cares if we send a success event, but it's probably good practice).

sleshchenko commented on brokers/artifacts/extensions.go:
```diff
@@ -31,20 +31,24 @@ func (b *Broker) ProcessPlugin(plugin *model.CachedPlugin) error {
 		if path != "" {
-			infoLine.WriteString(fmt.Sprintf("\n    Plugin already downloaded"))
+			logBuf = append(logBuf, fmt.Sprintf("    Plugin already downloaded"))
+			b.flushLog(logBuf)
```
When plugin is already downloaded iteration should be really fast, I guess we don't really need to flush buffer here.
But I'm OK with flushing it here if it does not lead to propagating logs in the wrong order.

amisevsk replied:
You're right, it was actually a mistake in `b.flushLog` that caused me to put this line here (the deferred flush gets the just-initialized buffer so it always logged nothing). This is fixed now.
